src/timely-util/src/operators_async_ext.rs

```diff
@@ -166,24 +178,49 @@ impl<G: Scope> OperatorBuilderExt<G> for OperatorBuilder<G> {
                 if had_pending_notify && !scheduler.is_notified() {
                     waker.wake_by_ref();
                 }
+
+                operator_incomplete
             }
         });
     }
 }
 
-#[allow(unused_macros)]
+use differential_dataflow::lattice::Lattice;
+#[doc(hidden)]
+pub fn multi_meet<T: Lattice + Clone>(antichains: &[Antichain<T>]) -> Antichain<T> {
+    antichains
+        .iter()
+        .fold(Antichain::new(), |acc, x| acc.meet(&x))
+}
+
+/// Convenience macro used to wrap what might otherwise be the argument to `build_reschedule`.
+#[macro_export]
 macro_rules! async_op {
-    (|$capabilities:ident, $frontiers:ident| $body:block) => {
-        move |mut capabilities, mut frontiers, scheduler| async move {
+    (|$capabilities:ident, $frontier:ident| $body:block) => {
+        move |mut capabilities, frontiers, scheduler| async move {
             loop {
                 scheduler.notified().await;
-                // rebind to mutable references to make sure they can't be accidentally dropped
+
+                // We simplify into one single Antichain because no current users of `async_op`
+                // require treating multiple input frontiers separately and this is the easiest
+                // way we can ensure that the `$body` does not hold a borrow to the frontier at
+                // all times
+                //
+                // Pull out to function so we don't require import from differential_dataflow for
+                // user of macro.
+                let __frontier = $crate::operators_async_ext::multi_meet(&frontiers.borrow());
```

Comment:
This looks scary, I'm not sure what the meet of the frontier of all inputs represents. If we want to give owned copies to `$body` we should just do `let $frontiers = (*frontiers.borrow()).clone()` to get a `Vec<Antichain<Timestamp>>` with one antichain per input. You could also pull the vector out of the loop to be reusing the allocation.

Reply:
I've removed it from this PR but, fwiw, we do functionally the same thing here when computing the `durability_frontier` for the kafka sink: https://github.com/MaterializeInc/materialize/blob/3d7db156adcfb0c7e37c6a02f1e41a2402739649/src/dataflow/src/sink/kafka.rs#L767